File: backend/app/core/categories.py
"""
Canonical Food Categories - Single Source of Truth

This file defines the official food categories for the Smart Canteen system.
All components must use these categories to ensure consistency.
"""

from typing import List, Dict
from enum import Enum

class CategoryKey(str, Enum):
    """Canonical category keys - use these for storage and logic"""
    MAIN_COURSE = "main_course"
    STARTER = "starter"
    SNACKS = "snacks"
    BEVERAGES = "beverages"
    DESSERT = "dessert"
    BREAD = "bread"
    RICE = "rice"
    SIDE_DISH = "side_dish"

# Canonical category definitions with display labels
CANONICAL_CATEGORIES: List[Dict[str, str]] = [
    {"key": CategoryKey.MAIN_COURSE, "label": "Main Course"},
    {"key": CategoryKey.STARTER, "label": "Starter"},
    {"key": CategoryKey.SNACKS, "label": "Snacks"},
    {"key": CategoryKey.BEVERAGES, "label": "Beverages"},
    {"key": CategoryKey.DESSERT, "label": "Dessert"},
    {"key": CategoryKey.BREAD, "label": "Bread"},
    {"key": CategoryKey.RICE, "label": "Rice"},
    {"key": CategoryKey.SIDE_DISH, "label": "Side Dish"},
]

# Helper functions for category management
def get_all_categories() -> List[Dict[str, str]]:
    """Returns all canonical categories"""
    return CANONICAL_CATEGORIES

def get_category_label(category_key: str) -> str:
    """Get display label for a category key"""
    for cat in CANONICAL_CATEGORIES:
        if cat["key"] == category_key:
            return cat["label"]
    return category_key.replace("_", " ").title()

def get_category_key(label: str) -> str:
    """Get category key from display label (case-insensitive)"""
    label_lower = label.lower()
    for cat in CANONICAL_CATEGORIES:
        if cat["label"].lower() == label_lower:
            return cat["key"]
    return label.lower().replace(" ", "_")

def is_valid_category(category_key: str) -> bool:
    """Check if category key is valid"""
    return category_key in [cat["key"] for cat in CANONICAL_CATEGORIES]

def get_category_keys() -> List[str]:
    """Get all category keys"""
    return [cat["key"] for cat in CANONICAL_CATEGORIES]

def get_category_labels() -> List[str]:
    """Get all category labels"""
    return [cat["label"] for cat in CANONICAL_CATEGORIES]

# Legacy category mapping for migration
LEGACY_CATEGORY_MAPPING = {
    "Main Course": CategoryKey.MAIN_COURSE,
    "Starter": CategoryKey.STARTER,
    "Snacks": CategoryKey.SNACKS,
    "Beverages": CategoryKey.BEVERAGES,
    "Beverage": CategoryKey.BEVERAGES,  # Handle singular form
    "Dessert": CategoryKey.DESSERT,
    "Desserts": CategoryKey.DESSERT,  # Handle plural form
    "Bread": CategoryKey.BREAD,
    "Rice": CategoryKey.RICE,
    "Side Dish": CategoryKey.SIDE_DISH,
    "Salad": CategoryKey.SIDE_DISH,  # Map Salad to Side Dish
    "Dietary": CategoryKey.SNACKS,  # Map Dietary to Snacks
    "Breakfast": CategoryKey.STARTER,  # Map Breakfast to Starter
    "Soup": CategoryKey.STARTER,  # Map Soup to Starter
}

def migrate_legacy_category(legacy_category: str) -> str:
    """Migrate legacy category name to canonical key"""
    if not legacy_category:
        return CategoryKey.MAIN_COURSE  # Default fallback
    
    # Direct mapping first
    if legacy_category in LEGACY_CATEGORY_MAPPING:
        return LEGACY_CATEGORY_MAPPING[legacy_category]
    
    # Try case-insensitive matching
    legacy_lower = legacy_category.lower()
    for legacy, canonical in LEGACY_CATEGORY_MAPPING.items():
        if legacy.lower() == legacy_lower:
            return canonical
    
    # Try to match with canonical labels
    return get_category_key(legacy_category)

File: frontend/src/core/categories.ts
/**
 * Canonical Food Categories - Frontend Types and Utilities
 * 
 * This file provides TypeScript types and utilities for the canonical
 * category system. It mirrors the backend category definitions.
 */

export interface Category {
  key: string;
  label: string;
}

export type CategoryKey = 
  | 'main_course'
  | 'starter' 
  | 'snacks'
  | 'beverages'
  | 'dessert'
  | 'bread'
  | 'rice'
  | 'side_dish';

// API endpoints for categories
export const CATEGORIES_API = {
  BASE: '/api/menu/categories',
  ALL: '/api/menu/categories/',
  KEYS: '/api/menu/categories/keys',
  LABELS: '/api/menu/categories/labels',
  VALIDATE: (key: string) => `/api/menu/categories/validate/${key}`,
  LABEL_FOR_KEY: (key: string) => `/api/menu/categories/label/${key}`,
  KEY_FOR_LABEL: (label: string) => `/api/menu/categories/key/${label}`,
};

// Utility functions
export const getCategoryLabel = (categories: Category[], key: string): string => {
  const category = categories.find(cat => cat.key === key);
  return category?.label || key.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
};

export const getCategoryKey = (categories: Category[], label: string): string => {
  const category = categories.find(cat => cat.label.toLowerCase() === label.toLowerCase());
  return category?.key || label.toLowerCase().replace(/\s+/g, '_');
};

export const isValidCategory = (categories: Category[], key: string): boolean => {
  return categories.some(cat => cat.key === key);
};

export const formatCategoryForDisplay = (key: string): string => {
  return key.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
};

// Sample fallback categories (used when API is unavailable)
export const FALLBACK_CATEGORIES: Category[] = [
  { key: 'main_course', label: 'Main Course' },
  { key: 'starter', label: 'Starter' },
  { key: 'snacks', label: 'Snacks' },
  { key: 'beverages', label: 'Beverages' },
  { key: 'dessert', label: 'Dessert' },
  { key: 'bread', label: 'Bread' },
  { key: 'rice', label: 'Rice' },
  { key: 'side_dish', label: 'Side Dish' },
];
